### ISSUE TITLE
Can't connect through the login form

### ISSUE THREAD

**Issue by bynicolas** [Adldap2/Adldap2-Laravel#771]
- Laravel Version: 5.8.32
- Adldap2-Laravel Version: 6.0.7
- PHP Version: 7.3.7
- LDAP Type: OpenLDAP

### Description:
Can't authenticate user from the login form, while being able to authenticate directly with `$provider->auth()->attempt('jennyfer', 'jen-j2019' )`.

So I'm sure the connection between the app and the server is fine, but clearly something is wrong or missing when trying to use the login form and I can't figure out what is. 

This is what I get in the logs

```
[2019-08-18 22:56:26] local.INFO: User 'einstein' has been successfully found for authentication.  
[2019-08-18 22:56:26] local.INFO: User 'einstein' is being synchronized.  
[2019-08-18 22:56:26] local.INFO: User 'einstein' has been successfully synchronized.  
[2019-08-18 22:56:26] local.INFO: User 'einstein' is authenticating with username: 'uid=,cn=users,dc=ldap,dc=example,dc=net'  
[2019-08-18 22:56:26] local.ERROR: A username must be specified. {"exception":"[object] (Adldap\\Auth\\UsernameRequiredException(code: 0): A username must be specified. at vendor/adldap2/adldap2/src/Auth/Guard.php:178)
```

If you note in the logs, you'll notice that the DN is missing the username value and I don't know why `'uid=[missing],cn=users,dc=ldap,dc=example,dc=net'`. I guess this is part of the problem.

Also, the logs says that the user is successfully synchronized but I don't see it saved in the database, is that normal behaviour?

#### ldap_auth.php
```
<?php

return [

    'connection' => env('LDAP_CONNECTION', 'default'),
    'provider' => Adldap\Laravel\Auth\DatabaseUserProvider::class,
    'model' => App\User::class,

    'rules' => [
        //
    ],

    'scopes' => [
        // 
    ],

    'identifiers' => [
        'ldap' => [
            'locate_users_by' => 'uid',
            'bind_users_by' => 'distinguishedname',
        ],

        'database' => [
            'guid_column' => 'objectguid',
            'username_column' => 'username',
        ],

        'windows' => [
            'locate_users_by' => 'samaccountname',
            'server_key' => 'AUTH_USER',
        ],
    ],

    'passwords' => [
        'sync' => env('LDAP_PASSWORD_SYNC', false),
        'column' => 'password',
    ],

    'login_fallback' => env('LDAP_LOGIN_FALLBACK', true),

    'sync_attributes' => [
        'username' => 'uid',
        'name' => 'cn',
    ],

    'logging' => [

        'enabled' => env('LDAP_LOGGING', true),
        'events' => [
            \Adldap\Laravel\Events\Importing::class                 => \Adldap\Laravel\Listeners\LogImport::class,
            \Adldap\Laravel\Events\Synchronized::class              => \Adldap\Laravel\Listeners\LogSynchronized::class,
            \Adldap\Laravel\Events\Synchronizing::class             => \Adldap\Laravel\Listeners\LogSynchronizing::class,
            \Adldap\Laravel\Events\Authenticated::class             => \Adldap\Laravel\Listeners\LogAuthenticated::class,
            \Adldap\Laravel\Events\Authenticating::class            => \Adldap\Laravel\Listeners\LogAuthentication::class,
            \Adldap\Laravel\Events\AuthenticationFailed::class      => \Adldap\Laravel\Listeners\LogAuthenticationFailure::class,
            \Adldap\Laravel\Events\AuthenticationRejected::class    => \Adldap\Laravel\Listeners\LogAuthenticationRejection::class,
            \Adldap\Laravel\Events\AuthenticationSuccessful::class  => \Adldap\Laravel\Listeners\LogAuthenticationSuccess::class,
            \Adldap\Laravel\Events\DiscoveredWithCredentials::class => \Adldap\Laravel\Listeners\LogDiscovery::class,
            \Adldap\Laravel\Events\AuthenticatedWithWindows::class  => \Adldap\Laravel\Listeners\LogWindowsAuth::class,
            \Adldap\Laravel\Events\AuthenticatedModelTrashed::class => \Adldap\Laravel\Listeners\LogTrashedModel::class,
        ],
    ],
];
```

#### ldap.php
```
<?php

return [

    'logging' => env('LDAP_LOGGING', false),

    'connections' => [

        'default' => [
            'auto_connect' => env('LDAP_AUTO_CONNECT', true),
            'connection' => Adldap\Connections\Ldap::class,

            'settings' => [
                'schema' => Adldap\Schemas\OpenLDAP::class,
                'account_prefix' => env('LDAP_ACCOUNT_PREFIX', '');
                'account_suffix' => env('LDAP_ACCOUNT_SUFFIX', ''),
                'hosts' => explode(' ', env('LDAP_HOSTS', 'corp-dc1.corp.acme.org corp-dc2.corp.acme.org')),
                'port' => env('LDAP_PORT', 389),
                'timeout' => env('LDAP_TIMEOUT', 5),
                'base_dn' => env('LDAP_BASE_DN', 'dc=corp,dc=acme,dc=org'),
                'username' => env('LDAP_USERNAME'),
                'password' => env('LDAP_PASSWORD'),
                'follow_referrals' => false,
                'use_ssl' => env('LDAP_USE_SSL', false),
                'use_tls' => env('LDAP_USE_TLS', false),
                'custom_options' => [
                    // See: http://php.net/ldap_set_option
                    LDAP_OPT_X_TLS_REQUIRE_CERT => LDAP_OPT_X_TLS_NEVER
                ]
            ],
        ],
    ],
];
```

#### .env
```
LDAP_HOSTS=10.0.1.5
LDAP_BASE_DN="dc=ldap,dc=example,dc=net"
LDAP_USERNAME="uid=root,cn=users,dc=ldap,dc=example,dc=net"
LDAP_PASSWORD=xxx
LDAP_ACCOUNT_PREFIX="uid="
LDAP_ACCOUNT_SUFFIX=",cn=users,dc=ldap,dc=example,dc=net"
LDAP_USE_SSL=true
LDAP_USE_TLS=false
LDAP_PORT=636
LDAP_LOGGING=true
```

**Comment by bynicolas** [comment]
Ok, so I've managed to make it work, but I still don't understand the issue.

I've narrowed it down to `LDAP_ACCOUNT_PREFIX`, `LDAP_ACCOUNT_SUFFIX` and `'bind_users_by' => 'distinguishedname',` being the issue.

I've changed `'bind_users_by' => 'distinguishedname',` to `'bind_users_by' => 'dn',` and removed the `LDAP_ACCOUNT_PREFIX` and `LDAP_ACCOUNT_SUFFIX` because I've noticed that when using `'bind_users_by' => 'dn',` the usersename used for authentication on the LDAP server was in the form of

```
'uid=uid=einstein,cn=users,dc=ldap,dc=example,dc=net,cn=users,dc=ldap,dc=example,dc=net'
```

So somehow, using `'bind_users_by' => 'dn',` already populates the correct prefix and suffix, so adding them from the config generates a wrong credential.

Maybe I've configured something wrong, so I still would like more context on why this is happening and how to fix it.

Thanks

**Comment by bynicolas** [comment]
Ok, I guess I've figured out what's happening, `dn` is being returned by the LDAP server after the `uid` has been filtered, and the `dn` already contains all the authentication syntax.

So if we want to use `'bind_users_by' => 'dn'`, then we cannot set the prefix or suffix (at least in my setup) as it will add unnecessary strings to the authenticating user.

If I am to use `LDAP_ACCOUNT_PREFIX` and `LDAP_ACCOUNT_SUFFIX`, then I need to set `'bind_users_by' => 'uid'` (again in my setup) to make it work.

I guess this is how it's supposed to work.

I'll close this issue since it's working great now, but I would still like some more context on the "issue" and a confirmation that what I've figured out is correct when you have a chance.
